### ISSUE TITLE
At the minute I use this slightly hacky repo in order to calculate the thresholds I want to deploy.

### ISSUE THREAD

**Issue by GeorgePearse** [roboflow/supervision#632]
### Search before asking

- [X] I have searched the Supervision [issues](https://github.com/roboflow/supervision/issues) and found no similar feature requests.

### Description

Create a simple API to find the best thresholds to maximise some metric (f1-score, precision, recall), given an annotated dataset and a model. 

https://github.com/yhsmiley/fdet-api

### Use case

Anyone wanting to deploy models without manual thresholding (or viewing graphs). 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by RigvedRocks** [comment]
I'd like to help by submitting a PR

**Comment by SkalskiP** [comment]
Hi @GeorgePearse and @RigvedRocks 👋🏻 ! Thanks for your interest in `supervision`. I am sorry that I have not been responsive for the last few days. Before Christmas, I was busy with duties unrelated to supervision, and I was off for the last few days.

The idea looks interesting. @RigvedRocks could you share some initial ideas regarding implementation?

**Comment by RigvedRocks** [comment]
I was thinking of using the basic techniques used in ML such as using the roc curve or using Youden J's statistic but the above approach outlined by @GeorgePearse works for me. I guess I can collaborate with @GeorgePearse to work on this issue if he insists so.

**Comment by GeorgePearse** [comment]
I'd really like to do what I can to keep this ticking over, @SkalskiP do you also think it's valuable? I'm always surprised by the lack of open-source implementations for this, and assume that every company just has their own fix. 

@RigvedRocks we could do something like I try to create a branch with a "workable" solution from fdet-api, but starting from the supervision format, and you take it from there? Let me know if that might interest you?

**Comment by SkalskiP** [comment]
Hi @GeorgePearse 👋🏻 I like the idea and I'd love to look at your initial implementation. If possible, I want the solution:
- As much as possible, used the Metrics API already found in Supervision
- If possible, it did not use external libraries. One of the main principles of Supervision is to limit external dependencies.

Such a solution requires a lot of steps, so I need to understand how we can combine it with what we have and how to design the next elements to be as reusable as possible.

**Comment by GeorgePearse** [comment]
Yeah all makes sense, tbh, the reason I want it to be integrated into supervision is to solve those very problems, at the minute I'm dealing with a lot of opaque code, I only trust the outputs from having visually inspected the predictions from lots of model/threshold combos that have used it. 

As for API questions. 

Just something like 

```python
# Ideally target_metric could also be a callback 
per_class_thresholds: dict = optimize_thresholds(
    predictions_dataset,
    annotations_dataset, 
    target_metric='f1_score', 
    per_class=True,
) 
```

**Comment by SkalskiP** [comment]
And what is stored inside `per_class_thresholds`? `Dict[int, float]` - class id to optimal IoU mapping?

What's inside `optimize_thresholds`? I'd appreciate any pseudocode.

**Comment by GeorgePearse** [comment]
class id to optimal score, the minimum IoU to classify a prediction and annotation as a match is set upfront by the user. Is that not the far more common use case for shipping ML products? The minimum IoU is defined by business/product requirements / can be done easily enough visually on a handful of examples. Maybe I'm biased by having mostly trained models where localisation is of secondary importance to classification, and a much much easier problem.

**Comment by GeorgePearse** [comment]
Complete pseudocode:

```python
metrics = []
for class_name in class_list: 
    for threshold in range(0, 1, 100): 
          current_metric = calculate_metric(grid_of_matched_predictions_and_their_scores, metric='f1_score')
          metrics.append({
              'threshold': threshold,
              'class_name': class_name,
              'metric': metric,
          })

metrics_df = pd.DataFrame(metrics)

# but so that you get a row per class, whatever the .groupby() kind of query 
# would be to achieve that 
best_metrics = metrics_df[metrics_df['metric'] == max(metrics_df['metric'])

```

But everything probably needs to calculated in numpy to not make it painfully slow. 

There's a decent chance that this is where most people get this data from currently https://github.com/rafaelpadilla/Object-Detection-Metrics, but the repo is as you'd expect of something 5/6 years old, and doesn't have the useability/documentation of a modern open-core project.

**Comment by SkalskiP** [comment]
Understood. This sounds interesting to me. I'm worried about scope, especially if we want to reimplement all metrics. 

- We need to divide work into smaller chunks. I am plodding with reviews when I need to go through 2k lines of code. On top of this we could assign different tasks to different external contributors and speed up the work. 
- We need to develop MVP - the shortest path demonstrating the value of the potential solution. With only 1 metric for example.

**Comment by RigvedRocks** [comment]
@GeorgePearse Fine by me. You can create a new branch called and then I can refine your initial solution.

**Comment by GeorgePearse** [comment]
From a look through the metric functionality already implemented it looks like it wouldn't be too painful to add in. The object that comes out of this looks like it's already done most of the upfront maths needed.

![image](https://github.com/roboflow/supervision/assets/47161914/30481339-66ac-4f86-b964-f6ec4c07baee)

**Comment by SkalskiP** [comment]
@GeorgePearse, could you be a bit more specific? What do you mean by scores?
